File: apps/basecamp/template/.claude/skills/sync-apple-calendar/scripts/sync.py
#!/usr/bin/env python3
"""Sync Apple Calendar events to ~/.cache/fit/basecamp/apple_calendar/ as JSON.

Queries the macOS Calendar SQLite database for events in a 14-day sliding
window (past and future) and writes one JSON file per event.

Usage: python3 scripts/sync.py

Requires: macOS with Calendar app configured and Full Disk Access granted.
"""

import json
import os
import subprocess
from datetime import datetime, timezone, timedelta

EPOCH = datetime(2001, 1, 1, tzinfo=timezone.utc)
OUTDIR = os.path.expanduser("~/.cache/fit/basecamp/apple_calendar")

DB_PATHS = [
    os.path.expanduser(
        "~/Library/Group Containers/group.com.apple.calendar/Calendar.sqlitedb"
    ),
    os.path.expanduser("~/Library/Calendars/Calendar.sqlitedb"),
]

STATUS_MAP = {
    0: "unknown",
    1: "pending",
    2: "accepted",
    3: "declined",
    4: "tentative",
    5: "delegated",
    6: "completed",
    7: "in-process",
}

ROLE_MAP = {0: "unknown", 1: "required", 2: "optional", 3: "chair"}


def find_db():
    db = next((p for p in DB_PATHS if os.path.exists(p)), None)
    if not db:
        print("Error: Apple Calendar database not found. Is Calendar configured?")
        exit(1)
    return db


def query(db, sql):
    result = subprocess.run(
        ["sqlite3", "-readonly", "-json", db, sql], capture_output=True, text=True
    )
    if result.returncode != 0:
        if "database is locked" in result.stderr:
            import time

            time.sleep(2)
            result = subprocess.run(
                ["sqlite3", "-readonly", "-json", db, sql],
                capture_output=True,
                text=True,
            )
        if result.returncode != 0:
            print(f"SQLite error: {result.stderr.strip()}")
            return []
    return json.loads(result.stdout) if result.stdout.strip() else []


def coredata_to_iso(ts, tz_name=None):
    """Convert Core Data timestamp to ISO 8601."""
    if ts is None:
        return None
    dt = EPOCH + timedelta(seconds=ts)
    if tz_name and tz_name != "_float":
        try:
            from zoneinfo import ZoneInfo

            dt = dt.astimezone(ZoneInfo(tz_name))
        except Exception:
            pass
    return dt.isoformat()


def main():
    db = find_db()
    os.makedirs(OUTDIR, exist_ok=True)

    now = datetime.now(timezone.utc)
    start = now - timedelta(days=14)
    end = now + timedelta(days=14)
    START_TS = (start - EPOCH).total_seconds()
    END_TS = (end - EPOCH).total_seconds()

    # Fetch events with a single query
    events = query(
        db,
        f"""
    SELECT
        ci.ROWID AS id,
        ci.summary,
        ci.start_date,
        ci.end_date,
        ci.start_tz,
        ci.end_tz,
        ci.all_day,
        ci.description,
        ci.has_attendees,
        ci.conference_url,
        loc.title AS location,
        cal.title AS calendar_name,
        org.address AS organizer_email,
        org.display_name AS organizer_name
    FROM CalendarItem ci
    LEFT JOIN Location loc ON loc.ROWID = ci.location_id
    LEFT JOIN Calendar cal ON cal.ROWID = ci.calendar_id
    LEFT JOIN Identity org ON org.ROWID = ci.organizer_id
    WHERE ci.start_date <= {END_TS}
        AND COALESCE(ci.end_date, ci.start_date) >= {START_TS}
        AND ci.summary IS NOT NULL
        AND ci.summary != ''
    ORDER BY ci.start_date ASC
    LIMIT 1000;
    """,
    )

    # Collect event IDs for batch attendee query
    event_ids = [str(ev["id"]) for ev in events]

    # Batch-fetch all attendees in one query (avoids N+1)
    attendees_by_event = {}
    if event_ids:
        id_list = ",".join(event_ids)
        attendees_raw = query(
            db,
            f"""
        SELECT
            p.owner_id,
            p.email,
            p.status,
            p.role,
            p.is_self,
            p.entity_type,
            i.display_name
        FROM Participant p
        LEFT JOIN Identity i ON i.ROWID = p.identity_id
        WHERE p.owner_id IN ({id_list})
            AND p.entity_type = 7;
        """,
        )
        for a in attendees_raw:
            oid = a["owner_id"]
            attendees_by_event.setdefault(oid, []).append(a)

    # Write event JSON files
    written_ids = set()
    for ev in events:
        eid = ev["id"]

        # Organizer — strip mailto: prefix from Identity.address
        org_email = ev.get("organizer_email") or None
        if org_email and org_email.startswith("mailto:"):
            org_email = org_email[7:]

        # Attendees
        attendees = []
        for a in attendees_by_event.get(eid, []):
            if not a.get("email"):
                continue
            attendees.append(
                {
                    "email": a["email"],
                    "name": (a.get("display_name") or "").strip() or None,
                    "status": STATUS_MAP.get(a.get("status"), "unknown"),
                    "role": ROLE_MAP.get(a.get("role"), "unknown"),
                    "self": bool(a.get("is_self")),
                }
            )

        is_all_day = bool(ev.get("all_day"))

        event_json = {
            "id": f"apple_cal_{eid}",
            "summary": ev["summary"],
            "start": {
                "dateTime": coredata_to_iso(ev["start_date"], ev.get("start_tz")),
                "timeZone": ev.get("start_tz")
                if ev.get("start_tz") != "_float"
                else None,
            },
            "end": {
                "dateTime": coredata_to_iso(
                    ev["end_date"] if ev["end_date"] else ev["start_date"],
                    ev.get("end_tz"),
                ),
                "timeZone": ev.get("end_tz")
                if ev.get("end_tz") != "_float"
                else None,
            },
            "allDay": is_all_day,
            "location": ev.get("location") or None,
            "description": ev.get("description") or None,
            "conferenceUrl": ev.get("conference_url") or None,
            "calendar": ev.get("calendar_name") or None,
            "organizer": {
                "email": org_email,
                "name": (ev.get("organizer_name") or "").strip() or None,
            }
            if org_email
            else None,
            "attendees": attendees if attendees else None,
        }

        filepath = os.path.join(OUTDIR, f"{eid}.json")
        with open(filepath, "w") as f:
            json.dump(event_json, f, indent=2)
        written_ids.add(f"{eid}.json")

    # Clean up events outside the window
    removed = 0
    for fname in os.listdir(OUTDIR):
        if fname.endswith(".json") and fname not in written_ids:
            os.remove(os.path.join(OUTDIR, fname))
            removed += 1

    print(f"Apple Calendar Sync Complete")
    print(f"Events synced: {len(written_ids)}")
    print(f"Time window: {start.date()} to {end.date()}")
    print(f"Files cleaned up: {removed} (outside window)")
    print(f"Output: {OUTDIR}")


if __name__ == "__main__":
    main()

File: products/basecamp/templates/.claude/skills/sync-apple-calendar/scripts/sync.mjs
#!/usr/bin/env bun
/**
 * Sync Apple Calendar events to ~/.cache/fit/basecamp/apple_calendar/ as JSON.
 *
 * Queries the macOS Calendar SQLite database (via node:sqlite) for events in a
 * sliding window — N days in the past through 14 days in the future. Writes one
 * JSON file per event and removes files for events that fall outside the window.
 * Attendee details (name, email, status, role) are batch-fetched and included.
 *
 * Requires macOS with Calendar app configured and Full Disk Access granted.
 */

if (process.argv.includes("-h") || process.argv.includes("--help")) {
  console.log(`sync-apple-calendar — sync calendar events to JSON

Usage: node scripts/sync.mjs [--days N] [-h|--help]

Options:
  --days N     Days back to sync (default: 30)
  -h, --help   Show this help message and exit

Requires macOS with Calendar configured and Full Disk Access granted.`);
  process.exit(0);
}

import { DatabaseSync } from "node:sqlite";
import {
  existsSync,
  mkdirSync,
  readdirSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";

const HOME = homedir();
const OUTDIR = join(HOME, ".cache/fit/basecamp/apple_calendar");

/** Core Data epoch: 2001-01-01T00:00:00Z */
const EPOCH_MS = Date.UTC(2001, 0, 1);

const DB_PATHS = [
  join(
    HOME,
    "Library/Group Containers/group.com.apple.calendar/Calendar.sqlitedb",
  ),
  join(HOME, "Library/Calendars/Calendar.sqlitedb"),
];

const STATUS_MAP = {
  0: "unknown",
  1: "pending",
  2: "accepted",
  3: "declined",
  4: "tentative",
  5: "delegated",
  6: "completed",
  7: "in-process",
};

const ROLE_MAP = {
  0: "unknown",
  1: "required",
  2: "optional",
  3: "chair",
};

/**
 * Find the Apple Calendar database.
 * @returns {string}
 */
function findDb() {
  const db = DB_PATHS.find((p) => existsSync(p));
  if (!db) {
    console.error(
      "Error: Apple Calendar database not found. Is Calendar configured?",
    );
    process.exit(1);
  }
  return db;
}

/**
 * Open the database in read-only mode with retry on lock.
 * @param {string} dbPath
 * @returns {import("node:sqlite").DatabaseSync}
 */
function openDb(dbPath) {
  try {
    return new DatabaseSync(dbPath, { readOnly: true });
  } catch (err) {
    if (err.message.includes("locked")) {
      Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 2000);
      return new DatabaseSync(dbPath, { readOnly: true });
    }
    throw err;
  }
}

/**
 * Execute a read-only query and return results.
 * @param {import("node:sqlite").DatabaseSync} db
 * @param {string} sql
 * @returns {Array<Record<string, any>>}
 */
function query(db, sql) {
  try {
    return db.prepare(sql).all();
  } catch (err) {
    console.error(`SQLite error: ${err.message}`);
    return [];
  }
}

/**
 * Convert Core Data timestamp to ISO 8601.
 * @param {number | null} ts - Seconds since 2001-01-01
 * @param {string | null} tzName
 * @returns {string | null}
 */
function coredataToIso(ts, tzName) {
  if (ts == null) return null;
  const ms = EPOCH_MS + ts * 1000;
  const dt = new Date(ms);

  if (tzName && tzName !== "_float") {
    try {
      return (
        dt.toLocaleString("sv-SE", { timeZone: tzName }).replace(" ", "T") +
        getUtcOffset(dt, tzName)
      );
    } catch {
      // Fall through to UTC
    }
  }
  return dt.toISOString();
}

/**
 * Get UTC offset string for a timezone at a given instant.
 * @param {Date} dt
 * @param {string} tzName
 * @returns {string} e.g. "+02:00" or "-05:00"
 */
function getUtcOffset(dt, tzName) {
  try {
    const parts = new Intl.DateTimeFormat("en-US", {
      timeZone: tzName,
      timeZoneName: "longOffset",
    }).formatToParts(dt);
    const tzPart = parts.find((p) => p.type === "timeZoneName");
    if (tzPart) {
      // "GMT+2" → "+02:00", "GMT-5:30" → "-05:30", "GMT" → "+00:00"
      // eslint-disable-next-line security/detect-unsafe-regex -- fixed pattern matching trusted Intl API output; no backtracking risk
      const match = tzPart.value.match(/GMT([+-])(\d{1,2})(?::(\d{2}))?/);
      if (match) {
        const sign = match[1];
        const hours = match[2].padStart(2, "0");
        const mins = (match[3] ?? "00").padStart(2, "0");
        return `${sign}${hours}:${mins}`;
      }
      return "+00:00"; // GMT with no offset
    }
  } catch {
    // Fallback
  }
  return "Z";
}

// --- Main ---

function fetchEvents(db, startTs, endTs) {
  return query(
    db,
    `
    SELECT
      ci.ROWID AS id,
      ci.summary,
      ci.start_date,
      ci.end_date,
      ci.start_tz,
      ci.end_tz,
      ci.all_day,
      ci.description,
      ci.has_attendees,
      ci.conference_url,
      loc.title AS location,
      cal.title AS calendar_name,
      org.address AS organizer_email,
      org.display_name AS organizer_name
    FROM CalendarItem ci
    LEFT JOIN Location loc ON loc.ROWID = ci.location_id
    LEFT JOIN Calendar cal ON cal.ROWID = ci.calendar_id
    LEFT JOIN Identity org ON org.ROWID = ci.organizer_id
    WHERE ci.start_date <= ${endTs}
      AND COALESCE(ci.end_date, ci.start_date) >= ${startTs}
      AND ci.summary IS NOT NULL
      AND ci.summary != ''
    ORDER BY ci.start_date ASC
    LIMIT 1000;
  `,
  );
}

function fetchAttendeesByEvent(db, eventIds) {
  const attendeesByEvent = {};
  if (eventIds.length === 0) return attendeesByEvent;
  const idList = eventIds.join(",");
  const attendeesRaw = query(
    db,
    `
    SELECT p.owner_id, p.email, p.status, p.role, p.is_self, p.entity_type, i.display_name
    FROM Participant p
    LEFT JOIN Identity i ON i.ROWID = p.identity_id
    WHERE p.owner_id IN (${idList}) AND p.entity_type = 7;
  `,
  );
  for (const a of attendeesRaw) {
    attendeesByEvent[a.owner_id] ??= [];
    attendeesByEvent[a.owner_id].push(a);
  }
  return attendeesByEvent;
}

function buildAttendeeList(rawAttendees) {
  const attendees = [];
  for (const a of rawAttendees) {
    if (!a.email) continue;
    attendees.push({
      email: a.email,
      name: (a.display_name ?? "").trim() || null,
      status: STATUS_MAP[a.status] ?? "unknown",
      role: ROLE_MAP[a.role] ?? "unknown",
      self: Boolean(a.is_self),
    });
  }
  return attendees.length > 0 ? attendees : null;
}

function buildOrganizer(ev) {
  let orgEmail = ev.organizer_email ?? null;
  if (orgEmail?.startsWith("mailto:")) orgEmail = orgEmail.slice(7);
  if (!orgEmail) return null;
  return { email: orgEmail, name: (ev.organizer_name ?? "").trim() || null };
}

function tzOrNull(tz) {
  return tz !== "_float" ? tz : null;
}

function buildEventJson(ev, attendeesByEvent) {
  return {
    id: `apple_cal_${ev.id}`,
    summary: ev.summary,
    start: {
      dateTime: coredataToIso(ev.start_date, ev.start_tz),
      timeZone: tzOrNull(ev.start_tz),
    },
    end: {
      dateTime: coredataToIso(ev.end_date ?? ev.start_date, ev.end_tz),
      timeZone: tzOrNull(ev.end_tz),
    },
    allDay: Boolean(ev.all_day),
    location: ev.location || null,
    description: ev.description || null,
    conferenceUrl: ev.conference_url || null,
    calendar: ev.calendar_name || null,
    organizer: buildOrganizer(ev),
    attendees: buildAttendeeList(attendeesByEvent[ev.id] ?? []),
  };
}

function writeEventFiles(events, attendeesByEvent) {
  const writtenIds = new Set();
  for (const ev of events) {
    const eventJson = buildEventJson(ev, attendeesByEvent);
    const filename = `${ev.id}.json`;
    writeFileSync(join(OUTDIR, filename), JSON.stringify(eventJson, null, 2));
    writtenIds.add(filename);
  }
  return writtenIds;
}

function cleanupStaleFiles(writtenIds) {
  let removed = 0;
  for (const fname of readdirSync(OUTDIR)) {
    if (fname.endsWith(".json") && !writtenIds.has(fname)) {
      unlinkSync(join(OUTDIR, fname));
      removed++;
    }
  }
  return removed;
}

function main() {
  let daysBack = 30;
  const daysIdx = process.argv.indexOf("--days");
  if (daysIdx !== -1 && process.argv[daysIdx + 1]) {
    daysBack = parseInt(process.argv[daysIdx + 1], 10);
  }

  const dbPath = findDb();
  mkdirSync(OUTDIR, { recursive: true });

  const now = Date.now();
  const start = new Date(now - daysBack * 86400000);
  const end = new Date(now + 14 * 86400000);
  const startTs = (start.getTime() - EPOCH_MS) / 1000;
  const endTs = (end.getTime() - EPOCH_MS) / 1000;

  const db = openDb(dbPath);

  try {
    const events = fetchEvents(db, startTs, endTs);
    const eventIds = events.map((ev) => String(ev.id));
    const attendeesByEvent = fetchAttendeesByEvent(db, eventIds);
    const writtenIds = writeEventFiles(events, attendeesByEvent);
    const removed = cleanupStaleFiles(writtenIds);

    console.log("Apple Calendar Sync Complete");
    console.log(`Events synced: ${writtenIds.size}`);
    console.log(
      `Time window: ${start.toISOString().slice(0, 10)} to ${end.toISOString().slice(0, 10)}`,
    );
    console.log(`Files cleaned up: ${removed} (outside window)`);
    console.log(`Output: ${OUTDIR}`);
  } finally {
    db.close();
  }
}

main();
